File: stepik-java-api/src/main/java/org/stepik/api/objects/ObjectsContainer.java
package org.stepik.api.objects;

import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * @author meanmail
 */
public abstract class ObjectsContainer<T> {
    private Meta meta;

    public int getCount() {
        return getItems().size();
    }

    public boolean isEmpty() {
        return getCount() == 0;
    }

    @NotNull
    public abstract List<T> getItems();

    @NotNull
    public Meta getMeta() {
        if (meta == null) {
            meta = new Meta();
            meta.setPage(1);
        }
        return meta;
    }

    @NotNull
    public abstract Class<T> getItemClass();

    public T get(int index) {
        return getItems().get(index);
    }

    public T getFirst() {
        return getItems().get(0);
    }
}

File: stepik-java-api/src/main/java/org/stepik/api/objects/ObjectsContainer.kt
package org.stepik.api.objects

import com.google.gson.annotations.JsonAdapter
import com.google.gson.annotations.SerializedName
import org.stepik.api.client.serialization.DefaultAsEmptyMetaAdapter
import org.stepik.api.client.serialization.DefaultJsonConverter
import java.util.*


abstract class ObjectsContainer<T> : Iterable<T> {
    @SerializedName("meta")
    @JsonAdapter(DefaultAsEmptyMetaAdapter::class, nullSafe = false)
    var meta: Meta = Meta()

    val count: Int
        inline get() = items.size

    val isEmpty: Boolean
        inline get() = items.isEmpty()

    val isNotEmpty: Boolean
        inline get() = items.isNotEmpty()

    abstract val items: MutableList<T>

    abstract val itemClass: Class<T>

    fun first(): T {
        return items.first()
    }

    fun firstOrDefault(default: T): T {
        return items.firstOrNull() ?: default
    }

    operator fun get(index: Int): T {
        return items[index]
    }

    override fun iterator(): Iterator<T> {
        return items.iterator()
    }

    override fun spliterator(): Spliterator<T> {
        return items.spliterator()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ObjectsContainer<*>

        if (meta != other.meta) return false
        if (items != other.items) return false
        if (itemClass != other.itemClass) return false

        return true
    }

    override fun hashCode(): Int {
        var result = meta.hashCode()
        result = 31 * result + items.hashCode()
        result = 31 * result + itemClass.hashCode()
        return result
    }

    override fun toString(): String {
        return DefaultJsonConverter.toJson(this, true)
    }
}
